qiskit/circuit/quantumcircuit.py

```diff
@@ -1802,17 +1809,18 @@ def depth(self) -> int:
         # We treat barriers or snapshots different as
         # They are transpiler and simulator directives.
         # The max stack height is the circuit depth.
+        whitelist = set(filter(filter_function, self._data))
```

Comment:
This broke a bunch of the tests, because not all instructions are hashable.

Reply:
oh my bad I assumed they were and didn't think about it. Thinking about it more there will probably be a similar issue with the lists of args too as they're not hashable either. My concern before with it as a list is we're basically making it O(n**2) (actually a bit worse because we add a list traversal for the lookup for each bit in the args lists of each instruction) when it was O(n) before.

Reply:
What about creating a function `is_directive` in the `Instruction` class that checks `_directive` and use directly `instr.is_directive()` inside the `for` loop? Or `whitelist` brings some benefits that I'm missing?

Reply:
not sure why it's not hashable. i thought python uses the `id()` of the object as hash. 
if this doesn't work i should probably revert to `list()`. I don't think the `depth()` method has ever really been a performance bottleneck in anything.

Reply:
I think this problem can be side-stepped by not creating a `whilelist`, and instead of `if (inst, qargs, cargs) in whitelist:`, use something like `if filter_function((instr, qargs, cargs)):`. There isn't much advantage to building the `whilelist` ahead of time (`filter_function` needs to be called once on every instruction in the circuit regardless) but the list lookup can become slow for large circuits, e.g.:

```
qc = qk.QuantumCircuit(1)
for _ in range(10000):
    qc.x(0)
qc.depth()
```

takes 5.7 s with the whitelist, and 24 ms without.

(Python does have a default `__hash__` function based on `id()`, but it's not used for mutable classes which define an `__eq__`. ref: https://stackoverflow.com/a/29434112 )

Reply:
thanks that was incredibly helpful!